tactic/interactive.lean

```diff
@@ -254,37 +254,198 @@ solve_by_elim_aux opt.discharger opt.restr_hyp_set opt.max_rep
+                                         (iff.to_eq iff.comm ) ),
+                add_edge r a' b' p',
+                return p'
+           | (`(%%a₀ → %%a₁), `(%%b₀ → %%b₁)) :=
+             if ¬ a₁.has_var ∧ ¬ b₁.has_var then
+             do p₀ ← symm_eq a₀ b₀,
+                p₁ ← symm_eq a₁ b₁,
+                p' ← mk_app `congr_arg [`(implies),p₀,p₁],
+                add_edge r a' b' p',
+                return p'
+             else unify a' b' >> add_refl r a' *> mk_mapp `rfl [none,a]
+           | (_, _) :=
+             (do guard $ a'.get_app_fn.is_constant ∧
+                         a'.get_app_fn.const_name = b'.get_app_fn.const_name,
+                 (a'',pa') ← add_symm_proof r a',
+                 guard $ a'' =ₐ b',
+                 pure pa' )
+           end,
+           p' ← mk_eq_trans pa p,
+           add_edge r a' b' p',
+           mk_eq_symm pb >>= mk_eq_trans p'
+
+meta def find_same_type (r : tauto_state) : expr → list expr → tactic (expr × expr)
+| e []         := failed
+| e (H :: Hs) :=
+  do t  ← infer_type H,
+     t' ← infer_type e,
+     (prod.mk H <$> symm_eq r e t) <|> find_same_type e Hs
+
+private meta def contra_p_not_p (r : tauto_state) : list expr → list expr → tactic unit
+| []         Hs := failed
+| (H1 :: Rs) Hs :=
+  do t ← (extract_opt_auto_param <$> infer_type H1) >>= whnf,
+     (do a   ← match_not t,
+         (H2,p)  ← find_same_type r a Hs,
+         H2 ← to_expr ``( (%%p).mpr %%H2 ),
+         tgt ← target,
+         pr  ← mk_app `absurd [tgt, H2, H1],
+         tactic.exact pr)
+     <|> contra_p_not_p Rs Hs
+
+meta def contradiction_with (r : tauto_state) : tactic unit :=
+contradiction <|>
+do tactic.try intro1,
+   ctx ← local_context,
+   contra_p_not_p r ctx ctx
+
+meta def contradiction_symm :=
+using_new_ref (native.rb_map.mk _ _) contradiction_with
+
+meta def assumption_with (r : tauto_state) : tactic unit :=
+do { ctx ← local_context,
+     t   ← target,
+     (H,p) ← find_same_type r t ctx,
+     mk_eq_mpr p H >>= tactic.exact }
+<|> fail "assumption tactic failed"
+
+meta def assumption_symm :=
+using_new_ref (native.rb_map.mk _ _) assumption_with
 
 /--
   `tautology` breaks down assumptions of the form `_ ∧ _`, `_ ∨ _`, `_ ↔ _` and `∃ _, _`
   and splits a goal of the form `_ ∧ _`, `_ ↔ _` or `∃ _, _` until it can be discharged
   using `reflexivity` or `solve_by_elim`
 -/
 meta def tautology : tactic unit :=
-repeat (do
-  gs ← get_goals,
-  () <$ tactic.intros;
-  de_morgan_hyps,
-  casesm (some ()) [``(_ ∧ _),``(_ ∨ _),``(_ ↔ _),``(Exists _),``(false)];
-  constructor_matching (some ()) [``(_ ∧ _),``(_ ↔ _),``(true)],
-  try (refine ``( or_iff_not_imp_left.mpr _)),
-  try (refine ``( or_iff_not_imp_right.mpr _)),
-  gs' ← get_goals,
-  guard (gs ≠ gs') ) ;
-repeat
-(reflexivity <|> solve_by_elim <|>
- constructor_matching none [``(_ ∧ _),``(_ ↔ _),``(Exists _)] ) ;
-done
+using_new_ref (expr_map.mk _) $ λ r,
+do try (contradiction_with r);
+   try (assumption_with r);
```

Comment:
Why `;` instead of `,` here?

Reply:
If either closes the goal, this stops the proof script.